torch_xla/csrc/tensor_util.cpp

```diff
@@ -608,21 +650,46 @@ void PopulateTensorBuffer(const at::Tensor& tensor,
 
 xla::ComputationClient::DataPtr TensorToXlaData(const at::Tensor& tensor,
                                                 const xla::Shape& shape,
-                                                const Device& device) {
-  auto populate_fn =
-      [&](const xla::ComputationClient::TensorSource& source_tensor,
-          void* dest_buffer, size_t dest_buffer_size) {
-        PopulateTensorBuffer(tensor, source_tensor.shape, dest_buffer,
-                             dest_buffer_size, device);
-      };
+                                                const Device& device,
+                                                bool transfer_async) {
+  XLA_TIMED("TensorToData");
+  if (transfer_async) {
+    std::shared_ptr<DataAsync> async = std::make_shared<DataAsync>();
+    auto populate_mwait = std::make_shared<xla::util::MultiWait>(1);
+    auto populate_fn =
+        [&](const xla::ComputationClient::TensorSource& source_tensor,
+            void* dest_buffer, size_t dest_buffer_size) {
+          PopulateTensorBuffer(tensor, source_tensor.shape, dest_buffer,
+                               dest_buffer_size, device);
+          populate_mwait->Done();
+        };
 
-  std::vector<xla::ComputationClient::TensorSource> source_tensors;
-  source_tensors.emplace_back(shape, device.ToString(), std::move(populate_fn));
+    async->source_tensors.emplace_back(shape, device.ToString(),
+                                       std::move(populate_fn));
+    TransferToServerAsync(async, {device.ToString()});
+    XLA_CHECK_EQ(async->async_datas.size(), 1);
```

Comment:
ditto

Reply:
We expect one async data to come out of this async call because this function only take one tensor. If we created an var called `num_tensor_upload` it might be misleading because we always expect one in this case.

Reply:
If this is always expected to be `1` without an exception, then the input parameter is redundant. Otherwise, resisting to hardcode numbers is my preferred way of passing information. I gave a suggestion on what `define` can look like in the previous comment.

Reply:
updated